src/structuring_element.jl

```diff
@@ -0,0 +1,356 @@
+3×3 Matrix{Bool}:
+ 1  1  0
+ 1  1  0
+ 0  0  0
+
+julia> se_offsets = strel(CartesianIndex, se_mask) # displacement offsets to its center point
+3-element Vector{CartesianIndex{2}}:
+ CartesianIndex(-1, -1)
+ CartesianIndex(0, -1)
+ CartesianIndex(-1, 0)
+
+julia> se = strel(Bool, se_offsets)
+3×3 BitMatrix:
+ 1  1  0
+ 1  1  0
+ 0  0  0
+```
+
+See also [`strel_diamond`](@ref) and [`strel_window`](@ref) for SE constructors for two
+special cases.
+"""
+function strel end
+
+strel(se) = strel(strel_type(se), se)
+
+# convenient user interface without exporting MorphologySE
+strel(::Type{ET}, se::AbstractArray) where {ET<:CartesianIndex} = strel(SEOffset{strel_ndims(se)}(), se)
+strel(::Type{ET}, se::AbstractArray) where {ET<:Bool} = strel(SEMask{strel_ndims(se)}(), se)
+
+# constructor for special SEs
+_strel_array(se::SEDiamond) = SEDiamondArray(se)
+_strel_array(se::SEWindow) = SEWindowArray(se)
+
+"""
+    strel_diamond(img, [dims]; r=1)
+    strel_diamond(size, [dims]; r=1)
+
+Construct the N-dimensional structuring element (SE) for a diamond shape. If image is
+provided, then `size=(3, 3, ...)` and `(1, 2, ..., N)` are the default values for `size` and
+`dims`.
+
+```jldoctest; setup=:(using ImageMorphology)
+julia> img = rand(64, 64);
+
+julia> se = strel_diamond(img)
+3×3 ImageMorphology.SEDiamondArray{2, 2, 0}:
+ 0  1  0
+ 1  1  1
+ 0  1  0
+
+julia> se = strel_diamond((3,3), (1,)) # 3×3 mask along dimension 1
+3×3 ImageMorphology.SEDiamondArray{2, 1, 1}:
+ 0  1  0
+ 0  1  0
+ 0  1  0
+
+julia> se = strel_diamond((3,5); r=2) # 3×5 mask with radius 2
+3×5 ImageMorphology.SEDiamondArray{2, 2, 0}:
+ 0  1  1  1  0
+ 1  1  1  1  1
+ 0  1  1  1  0
+```
+
+!!! note "specialization and performance"
+    The diamond shape `SEDiamond` is a special type for which many morphology algorithms may
+    provide much more efficient implementations. For this reason, if one tries to
+    collect an `SEDiamondArray` into other array types (e.g. `Array{Bool}` via `collect`),
+    then a significant performance drop is very likely to occur.
+
+See also [`strel`](@ref) and [`strel_window`](@ref).
+"""
+function strel_diamond(img::AbstractArray{T,N}, dims=coords_spatial(img); kw...) where {T,N}
+    return strel_diamond(ntuple(i->3, N), dims; kw...)
+end
+function strel_diamond(sz::Dims{N}, dims::Dims{K}=ntuple(identity, N); kw...) where {N,K}
+    return _strel_array(SEDiamond{N}(sz, dims; kw...))
+end
+
+"""
+    strel_window(img, [dims]; r=1)
+    strel_window(size, [dims]; r=1)
+
+Construct the N-dimensional structuring element (SE) with all elements in the local window
+connected. If image is provided, then `size=(3, 3, ...)` and `(1, 2, ..., N)` are the
+default values for `size` and `dims`.
+
+```jldoctest; setup=:(using ImageMorphology)
+julia> img = rand(64, 64);
+
+julia> se = strel_window(img)
+3×3 ImageMorphology.SEWindowArray{2, 2, 0}:
+ 1  1  1
+ 1  1  1
+ 1  1  1
+
+julia> se = strel_window((3,3), (1,)) # 3×3 mask along dimension 1
```

Comment:
Personally I'd rather think of this as
```
1
1
1
```
and specify it as `strel_window((3,1))`. Or in terms of half-sizes, `strel_window((1,0))`.

Reply:
I would think both are equivalent representations so I won't be too worried about it. It totally depends on how the user creates it.

Actually, if we play with the `strel` conversion, we'll see they're eventually unified to the same:

```julia
julia> strel(Bool, strel(CartesianIndex, strel_box((3,1))))
3×1 BitMatrix:
 1
 1
 1

julia> strel(Bool, strel(CartesianIndex, strel_box((3,3), (1,))))
3×1 BitMatrix:
 1
 1
 1
```

But I'll make the default `strel_box(img, dims)` to use the smaller version:

```diff
  function strel_box(img::AbstractArray{T,N}, dims=coords_spatial(img); kw...) where {T,N}
-     return strel_box(ntuple(i->3, N), dims; kw...)
+     return strel_box(ntuple(i->in(i, dims) ? 3 : 1, N), dims; kw...)
  end
```